make default records fields and missing pb fields equivalent.

evanmcc opened the pull request:
Currently when generating using the proto3 syntax, `#foo{} =/= foo:decode(<<>>)`, which is kind of painful, especially when testing.  This PR makes it such that default fields are no longer undefined, but rather set to their type default, which should make them more like the output of decode.  Sub-messages are left as undefined; I'm not sure what the specification suggests, but I went with this behavior to match that of decode.

lukebakken commented:
If I'm reading the [`README`](https://github.com/tomas-abrahamsson/gpb#for-proto3-syntax) correctly, this should already be the behavior, so good find.

Keeping [`undefined` for sub-messages](https://developers.google.com/protocol-buffers/docs/proto3#default) appears to be the correct behavior.

FWIW, @tomas-abrahamsson did a lot of work related to this in proto 2 here: #82

evanmcc commented:
I'm just now realizing that the change of default type value for string (to binary, rather than character list strings) might not be what everyone wants!  I can revert it and change all of our usages to bytes, depending on what people think.

lukebakken commented:
It should probably depend on the [`strings_as_binaries` option](https://github.com/tomas-abrahamsson/gpb#mapping-of-protocol-buffer-datatypes-to-erlang).

evanmcc commented:
You're right.  I'll work on fixing that.

evanmcc commented:
OK, reverted the global change and tested that the `strings_as_binaries` option was doing the right thing for our defaults.

tomas-abrahamsson commented:
Thanks! I'll take a look. I think this is an ok addition.

> `#foo{} =/= foo:decode(<<>>)`

@evanmcc Could you elaborate a bit about the situation, why this property should hold? (It seems to me a bit of a special case)

> Keeping undefined for sub-messages appears to be the correct behavior.

@lukebakken I agree on this.

evanmcc commented:
Mostly for ease of testing.  We had a large existing base of tests that did things like:

```erlang
?assertMatch(#foo{field_3 = Something, field_7 = SomethingElse}, foo:decode(Bin))
```

Making the defaults of record construction match the ones from decode made the most sense to me, rather than fixing all the tests to check the fields individually.  Since defaults are ignored on encoding I figured it shouldn't have any broader effects.

tomas-abrahamsson commented:
Ok, I see.

One can still check several fields at once, no need to check them individually. Assume we have a `-record(xyz,{a,b,c})`, then this matches: `#xyz{a=1,c=3} = #xyz{a=1,b=false,c=3}`, but they are not equal. I guess some of the tests might perhaps do `?assertEqual(Term1,Term2)` instead of `?assertMatch(Pattern,Expr)`, cf `==` vs `=`. Could that be it?

tomas-abrahamsson commented:
I have a few comments on the code, will proceed to that.

tomas-abrahamsson commented on src/gpb_compile.erl:
```diff
@@ -6156,6 +6156,11 @@ format_msg_record(Msg, Fields, Opts, Defs) ->
+        case proplists:get_value(syntax, Defs, no) of
+            "proto3" -> true;
+            _ -> false
+        end,
```
Use this instead:
```erlang
    IsProto3 = gpb:is_msg_proto3(MsgName, Defs),
```
In a translation unit, it is possible to have a mix of proto2 and proto3 messages. All the parsed items for the .proto and all `import`ed files are concatenated, so there can actually be several `{syntax,...}` items in `Defs`. The first found `syntax` item may be the wrong one.

This applies to one more place below

evanmcc replied:
I can't do this at either location; MsgName isn't visible in either place.

tomas-abrahamsson replied:
You need to pass it in, it is available at call sites.

tomas-abrahamsson commented on src/gpb_compile.erl:
```diff
@@ -6183,20 +6188,28 @@ format_hfields(Indent, Fields, Opts, Defs) ->
+                                            io_lib:format(" = ~p", [Default]);
+                                        false -> ""
+                                        end;
+                                _ -> ""
```
Long line > 80: The `Default = proto3_type_default(...)` line. I'm a bit picky about that :) (I usually run with a narrow editor, 80 cols, but several such windows, so I can see more code at once, but overlong lines wrap and messes up reading...)

Minor: Please use the `?f/2` macro instead of io_lib:format

Minor: odd indentation of the `end` of `case IsProto3 of ...`

tomas-abrahamsson commented on src/gpb_compile.erl:
```diff
@@ -6318,7 +6331,11 @@ can_specify_map_item_presence_in_typespecs(Opts) ->
+                          case proplists:get_value(syntax, Defs, yes) of
+                              "proto3" -> "";
+                              _ -> " | undefined"
+                          end;
```
Do:
```erlang
    IsProto3 = gpb:is_msg_proto3(MsgName, Defs),
```
same as above.

tomas-abrahamsson replied:
ditto here, if you pass it in. It will be available at the call sites once you've passed it in to the location above

tomas-abrahamsson commented on src/gpb_compile.erl:
```diff
@@ -8489,7 +8506,7 @@ type_default(Type, Defs, Opts, GetTypeDefault) ->
             case get_strings_as_binaries_by_opts(Opts) of
                 true ->
-                    list_to_binary(GetTypeDefault(Type, Defs));
+                    iolist_to_binary(GetTypeDefault(Type, Defs));
```
Minor: Why change from `list_to_binary` into `iolist_to_binary`?
The default is never a binary here.
A reminiscence from the previous `strings_as_binaries` issue?

evanmcc replied:
No idea.  I may have noticed that it's the only usage of `list_to_binary` in the file:
```
$ git grep -c iolist_to_binary src/gpb_compile.erl
src/gpb_compile.erl:9
$ git grep -c list_to_binary src/gpb_compile.erl
src/gpb_compile.erl:10
```
but otherwise have no recollection.  I'll revert.

tomas-abrahamsson replied:
No problem, either way will work :)

tomas-abrahamsson commented:
Sorry, didn't notice you had updated the branch. Github does not send mails when only the branch was updated, so I failed to notice it. Looks good, will merge.

tomas-abrahamsson commented:
Hmm, having second thoughts about this: the `| undefined` in the type spec for proto3 messages has was dropped, but for encoding, the gpb still generates code like this:
```erlang
    B1 = if F1 == undefined -> Bin;
	    true -> <encoding work for field F1...>
	 end,
```
This was "legal" previously (due to `| undefined`), but not any more, and dialyzer (rightfully) complains about the generated code:
```
p23misc.erl:90: Guard test F1::integer() == 'undefined' can never succeed
```
I realize now this means two things:

1. The encoder in [`gpb_compile:field_encode_expr`](https://github.com/tomas-abrahamsson/gpb/blob/3.26.5/src/gpb_compile.erl#L2639L2686) needs to be more complex so that it does not generate code that dialyzer would flag as invalid: it will have to check for `undefined` only when the field type *can* be `undefined`, such as for sub message fields, but not for scalars.
2. Depending on how we do it, it might also mean different or deviating semantics for the case of maps with `maps_unset_optional=present_undefined`:
   * for maps, `undefined` is legal, and means "not set", but this now deviates from records. (This is how the PR currently does it) 
   * for maps, `undefined` is illegal, all values need to be set to defaults (`0` for integers and so on). Having it this way would make it work same as for records (no deviation), but it would be a non-backwards compatible change (differing semantics).

I'm not entirely happy with either of the implications. At a minimum, we could put back the `| undefined` part of the type spec.

Have you had a look at whether it would be possible to change your tests to do matching as opposed to checking for equality? Maybe that's really a better way to go?

evanmcc commented:
Sorry, I got pulled off onto another project briefly.

I don't understand your suggestion, they neither match nor are equal.  Your statement in: https://github.com/tomas-abrahamsson/gpb/pull/97#issuecomment-264026011 doesn't seem to be correct, `undefined` isn't `_`, is it?  I spent some time digging into the macros and trying out various things at the shell, but was unable to make it work.

tomas-abrahamsson commented:
Sorry, for not being very clear. I was thinking like this:

Even _without_ this PR, it should really already work the way you [described:](97#issuecomment-263988455)
> ```?assertMatch(#foo{field_3 = Something, field_7 = SomethingElse}, foo:decode(Bin))```

but I got the impression such tests failed, and that would make only sense if the the tests actually use `?assertEqual` instead of `?assertMatch`.

Suppose we have `x.proto`:
```proto
syntax="proto3";
message m { uint32 f = 1; uint32 g = 2; bool b = 3; };
```
This is in the Erlang shell, so I'm `=` instead of `?assertMatch` and `=:=` instead of `?assertEqual`:
```erlang
1> gpb:version_as_string().
"3.26.5" % currently running with plain gpb, PR not merged
2> rr("x.hrl").
[m]
3> #m{}. % m has 3 fields, we'll be checking only a subset of them
#m{f = undefined,g = undefined,b = undefined}
4> #m{f=0, g=0} = x:decode_msg(<<>>, m). % matches, so using ?assertMatch would succeed
#m{f = 0,g = 0,b = false}
5> #m{f=0, g=0} =:= x:decode_msg(<<>>, m). % not equal, so using ?assertEqual would fail
false
```

So I was speculating that an alternative to merging this PR could be to look over the tests, to change to `?assertMatch`. That was [my suggestion](https://github.com/tomas-abrahamsson/gpb/pull/97#issuecomment-264026011). This is of course just guessing, since I don't know what code base it is. (Is it available somewhere, would it be possible to take a look?) Do you follow my line of reasoning? I feel I might well have misunderstood something, please correct where I'm wrong.

tomas-abrahamsson commented:
Even with the latest fix (evanmcc/gpb@fda50b5), I get a dialyzer warning for the generated code. Example:
```protobuf
syntax="proto3";
message Test1 {
  int32    o_i32     = 11;
};
```
Excerpt from the generated `.hrl` file:
```erlang
-record('Test1',
        {o_i32 = 0              :: integer()        % = 11, 32 bits
        }).
```
Excerpt from the generated `.erl` file (with line numbers):
```erlang
    56  e_msg_Test1(#'Test1'{o_i32 = F1}, Bin, TrUserData) ->
    57      if F1 == undefined -> Bin;
    58         true ->
    59             begin
    60               TrF1 = id(F1, TrUserData),
    61               if TrF1 =:= 0 -> Bin;
    62                  true -> e_type_int32(TrF1, <<Bin/binary, 88>>)
    63               end
    64             end
    65      end.
```
Running dialyzer:
```bash
$ erl -pa ebin -noinput +B -boot start_clean -I`pwd` -o `pwd` -type -s gpb_compile c x2.proto
$ erlc -Wall +debug_info -Iinclude x2.erl
$ dialyzer -q --plt .gpb-19.2.1.plt  x2.beam
x2.erl:57: Guard test F1::integer() == 'undefined' can never succeed
x2.erl:188: Guard test NFo_i32::integer() =:= 'undefined' can never succeed
```
This is because in the record definition, `undefined` is not included in the field's type, but the generated erl code tests for it.

evanmcc commented:
Sorry, Tomas.  I've been pretty busy at work and haven't even had a chance
to work on addressing your comments.  I'll do my best to do so this week,
or at least explicate our testing issues in greater length.  I did this to
fix an internal dialyzer issue (we ignore the dialyzer output of the
generated code to save analysis time, sadly).  Sorry I've been
uncommunicative, I'll do my best to get back to this soon.

tomas-abrahamsson commented:
No problem, I've been a bit slow to respond myself, didn't know if you were expecting something from me or not, so I thought I'd better break the silence. No hurry.

tsloughter commented:
@tomas-abrahamsson is the PR acceptable if we simply keep `| undefined` for each field typespec?

tomas-abrahamsson commented:
I'll take a look, within a few days

tomas-abrahamsson commented:
I just pushed a commit that adds `| undefined`, and another that adds unit tests. If you are ok with those changes, please squash them and I think it's good for merge.

(Maybe in the long run it might be cleaner to drop the `| undefined`, and rewrite the encoder a bit with dialyzer's findings, but that could be a later step.)

Sorry for having been a bit back and forth about this, changing my mind and taking long time.

tsloughter commented:
Thanks! I've squashed the commits.

tomas-abrahamsson commented:
Included in 3.27.6, which was just pushed, hence closing this.
